text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,313 @@
+remains unchanged as we will store scale encoded `UMPSignal` messages directly in the parachain 
+UMP queue by outputing them in [upward_messages](https://github.com/paritytech/polkadot-sdk/blob/b5029eb4fd6c7ffd8164b2fe12b71bad0c59c9f2/polkadot/primitives/src/v7/mod.rs#L684). 
+
+The UMP queue layout is changed to allow the relay chain to receive both the XCM messages and 
+`UMPSignal` messages. An empty message (empty `Vec<u8>`) is used to mark the end XCM messages and 
+the start of `UMPSignal` messages.
+
+This way of representing the new messages has been chosen over introducing an enum wrapper to 
+minimize breaking changes of XCM message decoding in tools like Subscan for example. 
+
+Example: 
+```rust
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::SelectCore ]
+```
+
+#### `UMPSignal` messages
+
+```rust
+/// An `u8` wrap around sequence number. Typically this would be the least significant byte of the 
+/// parachain block number.
+pub struct CoreSelector(pub u8);
+
+/// An offset in the relay chain claim queue.
+pub struct ClaimQueueOffset(pub u8);
+
+/// Signals sent by a parachain to the relay chain.
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is committed to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and `ClaimQueueOffset`
+	/// to compute the index of the core the candidate has committed to.
+	///
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+```
+
+The parachain runtime is not concerned with the actual `CoreIndex` the candidate is intended for, 
+but must provide enough information for the validators and collators to compute it. `CoreSelector` 
+and `ClaimQueueOffset` use only 2 bytes and fulfill the requirement.
+
+**Example:**
+
+`cq_offset = 1` and `core_selector = 3`
+
+The table below represents a snapshot of the claim queue:
+
+|  | offset = 0 | offset = 1 | offset = 2 |
+| :--:	 | :--:   | :--:   | :--:   |
+| Core 1    | **Para A**     | **Para A**     | **Para A**     |
+| Core 2   | **Para A**  | Para B  | **Para A**  |
+| Core 3   | Para B  | **Para A**  | **Para A** |
+
+The purpose of `ClaimQueueOffset` is to select the column from the above table. 
+For `cq_offset = 1` we get `[ Para A, Para B, Para A]` and use as input to create 
+a sorted vec with the cores A is assigned to: `[ Core 1, Core 3]` and call it `para_assigned_cores`.
+We use `core_selector` and determine the committed core index is `Core 3` like this:
+
+```rust
+let committed_core_index = para_assigned_cores[core_selector % para_assigned_cores.len()];
+```
+
+Parachains should prefer to have a static `ClaimQueueOffset` value that makes sense for their 
+usecase which can be changed by governance at some future point. Changing the value dynamically 
+can be a friction point. It will work out fine to decrease the value to build more into the 
+present. But if the value is increased to build more into the future, a relay chain block will 
+be skipped.
```

Comment:
Maybe we should just remove this paragraph as it is mostly causing confusion.

Note also that this "friction" point has nothing to do with using a claim queue offset or some other solution - the friction of switching is fundamental and can not be mitigated by replacing `ClaimQueueOffset` with something else.

Let's iterate the two situations:

1. You decide to provide collations a bit more ahead of time, e.g. for getting backed in 2 blocks from now, instead of just one. You can do that by producing candidates faster at switch over time, sequence:

a) old settings, you produced a candidate because you saw your assignment upcoming in position 1.
b) settings switch, you are now aiming for (2).
c) you immediately produce another candidate (assuming core is assigned to you at position (2)). Basically if the core is assigned to your para at postion (1) and (2) you would produce two blocks in a row, without any delay at switch over time.

Situation two is the reverse:

(a) You build your candidate when it appeared at position (2) (old settings).
(b) setting switch, we are now aiming for (1). Now assuming we had another assignment at (3) earlier, we will now wait 12 seconds until it moved to (1) until we produced another block.


In fact there is not even too much friction. The only thing that happens is that parachain block times will have a "jump" at switch over time. In case (1), this might lead to one block being less filled for example. 


The real friction @sandreim was referring to has actually nothing to do with the claim queue itself, but rather with the relay parent advancement rule: If you move claim queue position from (1) to (2) because you want to build on the parent of the most current leaf instead of the leaf itself, in order to be more resilient to relay chain forks you will need to skip a relay chain block for the adjustment, because the relay parent block number most not be less than the previous one.

Reply:
Going to keep just first part about preferring a static offset. It's up to the implementation to figure out a good default value and documentation to give details.